staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -114,6 +123,32 @@ func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		}()
 		t.handler.ServeHTTP(w, rCopy)
 	}()
+
+	// Normally, the handler should notice the cancellation of the request
+	// context and stop. We give it some additional time and only then take
+	// over.
+	timeoutCh := make(chan struct{})
+
+	// This signals when ServeHTTP has returned, *not* when it is time to
+	// return. I.e. this intentionally doesn't use r.Context()!  We could
+	// wait for the timeout, but then the goroutine keeps running longer
+	// than it needs to.
+	done, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		defer close(timeoutCh)
+
+		// This context will get cancelled either in case of a timeout or when
+		// ServeHTTP returns.
+		<-r.Context().Done()
+
+		// Now wait a bit longer or until ServeHTTP returns.
+		select {
+		case <-time.After(timeoutFallbackDelay):
+		case <-done.Done():
+		}
+	}()
```

Comment:
this is *really* twisty... the timeout filter is already really difficult to reason about and this makes that aspect worse :-/

Reply:
is the race happening inside the `t.handler.ServeHTTP(w, rCopy)` call, or once this ServeHTTP method returns?

Reply:
The race occurs when `timeoutHandler.ServeHTTP` returns while the handler in the goroutine is still running.

Reply:
that seems like it can only happen if we receive from resultCh or timeoutCh (which is currently the context's Done channel)... I don't understand why we're decoupling from the context Done channel here... that seems like going in the wrong direction for this filter

Reply:
That was based on the suggestion from @lavalamp that once the request has timed out, the real handler should get some time to abort on its own before activating the fallback timeout code. The `timeoutCh` therefore does not get closed at the same time as the request cancellation, but `timeoutFallbackDelay` later.

Reply:
I was trying to suggest the other order, I think -- this horrible cleanup should trigger a couple seconds after the context has timed out, so that it only races with misbehaved handlers instead of all handlers.